scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -920,6 +922,68 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
     _.pApply(Sample.fixedSizeGlobally(sampleSize)).map(_.asScala)
   }
 
+  final private class WeightedHeapAggregator(totalWeight: Long, cost: T => Long)(implicit
+    ordering: Ordering[(Long, T)]
+  ) extends Aggregator[T, mutable.PriorityQueue[(Long, T)], Iterable[T]] {
+    @transient private lazy val random = new Random()
+
+    override def prepare(value: T): mutable.PriorityQueue[(Long, T)] = {
+      val queue = mutable.PriorityQueue[(Long, T)]()
+      if (cost(value) <= totalWeight) {
+        queue.enqueue((random.nextLong(), value))
+      }
+      queue
+    }
+
+    override def semigroup: Semigroup[mutable.PriorityQueue[(Long, T)]] = Semigroup.from {
+      (left, right) =>
+        var weight = 0L
+        val queue = mutable.PriorityQueue[(Long, T)]()
```

Comment:
This feels overcomplicated; shouldn't we just start w/ the most-full of the queues and add all the items from the other one, letting the queue evict the lesser-weighted elements?

Reply:
The implementation is based on a binary tree heap, where removal of lowest priority is costly. To achieve good result we should go with a [fibonnacci heap](https://en.wikipedia.org/wiki/Fibonacci_heap) priority queue, where merge and min removals are cheap.

I'm also annoyed on how algebird forces us to `prepare` each element in a mergable structure. I'm thinking of changing for `combine`, so we ca reuse the priority queue

Reply:
not the `lesser-weighted` but the lower priority yes

Reply:
Actually we can achieve this with a normal priority queue, with min heap.